File: iteration2/Transcript.java
import java.util.ArrayList;
import java.util.List;

//Transcript class for each student
//Keeps course list and grade list
public class Transcript {
    private List<Course> listOfCourses;
    private List<Grade> listOfGrades;
    private int studentCredits;
    private int studentYear;
    private double GPA;

    
    public Transcript() {

    }
    
    //Copy constructor,
    //Constructor to be used with an already existing Transcript Object
    public Transcript(Transcript transcript) {
        this.listOfCourses = transcript.listOfCourses;
        this.listOfGrades = transcript.listOfGrades;
        this.studentCredits = transcript.studentCredits;
    }
    //Constructor with parameters
    public Transcript(List<Course> listOfCourses, List<Grade> listOfGrades) {
        this.listOfCourses = listOfCourses;
        this.listOfGrades = listOfGrades;
        this.studentCredits = calculateCredit();
        this.GPA = calculateGPA();
        this.studentYear = calculateYear();
    }

    
    public boolean checkPassedCourses(Course checkedCourse) {
        boolean isAvailable = true;
        int i = 0;
        for (Course currentCourse1 : listOfCourses) {
            if (currentCourse1.equals(checkedCourse)) {
                //if it finds the course then checks whether it is retakable or not then it returns false or true immediately
                isAvailable = listOfGrades.get(i).isRetakableGrade();
                break;
            }
            i++;
        }
        return isAvailable;
    }

    private int calculateYear() {
        int year = 1;
        if (studentCredits >= 180) {
            year = 4;
        } else if (studentCredits >= 120) {
            year = 3;
        } else if (studentCredits >= 60) {
            year = 2;
        }
        if (year != 4 && GPA >= 3) {
            year++;
        }
        return year;
    }


    public int getYear() {
        return this.studentYear;
    }

    public int calculateCredit() throws ArrayIndexOutOfBoundsException {
        int totalCredit = 0;

        if (listOfCourses.isEmpty()) {
            throw new ArrayIndexOutOfBoundsException("No Courses For Student Found!");
        }
        int i = 0;
        for (Course listOfCourse : listOfCourses) {
            if (listOfGrades.get(i).getNumericalGrade() >= 35) {
                totalCredit += listOfCourse.getCourseCredit();
            }
            i++;

        }
        return totalCredit;
    }


    //Method to calculate current gpa
    public float calculateGPA() throws ArrayIndexOutOfBoundsException {
        float gpa;
        if (listOfCourses.isEmpty()) {
            throw new ArrayIndexOutOfBoundsException("No Courses For Student Found!");
        }
        if (listOfGrades.isEmpty()) {
            throw new ArrayIndexOutOfBoundsException("No Grades For Student Found!");
        }


        List<Integer> weightedValues = new ArrayList<>();
        int totalCredit = 0;
        //Calculate Weighted Values
        for (int i = 0; i < listOfCourses.size(); i++) {
            int credit = listOfCourses.get(i).getCourseCredit();
            int grade = listOfGrades.get(i).getNumericalGrade();
            weightedValues.add(credit * grade);
            totalCredit += credit;
        }

        int totalWeightedValues = 0;
        for (int i = 0; i < listOfCourses.size(); i++) {
            totalWeightedValues += weightedValues.get(i);
        }

        //calculate gpa 
        gpa = (float) (totalWeightedValues) / ((float) (totalCredit) * 25);

        return gpa;
    }

    public List<Course> getListOfCourses() {
        return listOfCourses;
    }
    public List<Grade> getListOfGrades() {
        return listOfGrades;
    }

    //calculate semester depending on the current credit
    //RIGHT NOW USES A PREDEFINED VALUE FOR EACH SEMESTERS WORTH!!
    //Moved to calculateYear()
    /*  public int calculateSemesterFromCredit() {
        int creditPreSemester = 3, semesterLimit = 16;
        int semester = 1;
        int completedCredits = this.calculateCredit();


        //used for instead of switch for future compatibility
        for (int i = 0; i < semesterLimit; i++) {
            if (completedCredits >= creditPreSemester) {
                semester++;
                completedCredits -= creditPreSemester;
            }
        }

        return semester;
    } */

    @Override
    public String toString() {
        String s = null;
        try {
            float gpa = this.calculateGPA();
            int credits = this.calculateCredit();
            int year = this.calculateYear();


            //adjust digit precision
            int digitPrecision = 2;
            int temp = (int) (gpa * Math.pow(10, digitPrecision));
            gpa = (float) ((float) temp / Math.pow(10, digitPrecision));


            s = "Cumulative Gpa: " + gpa + "\nCumulative Credits: " + credits +
                    "\nCurrent Year: " + year;
        } catch (ArrayIndexOutOfBoundsException e) {
            System.out.println(e.toString());
        }
        return s;
    }

    public String[] getStudentTranscriptStringList() {
        int courseCount = listOfGrades.size();
        String[] transcriptString = new String[courseCount + 1];
        transcriptString[0] = this.toString();
        for (int i = 1; i < courseCount + 1; i++) {
            transcriptString[i] = i + "-)" + listOfCourses.get(i - 1).toString() + " " + listOfGrades.get(i - 1).toString();
        }
        return transcriptString;
    }
}

File: iteration3/Transcript.py
import logging
from Course import Course


class Transcript:
    __list_of_courses = []
    __list_of_grades = []
    __student_credits: int = 0
    __student_year: int = 1
    __gpa: float = 0

    def __init__(self, list_of_courses, list_of_grades):
        self.__list_of_courses = list_of_courses
        self.__list_of_grades = list_of_grades
        self.__calculate_credit()
        self.__calculate_gpa()
        self.__calculate_year()

    # checks if a course is passed
    def check_passed_courses(self, checked_course: Course) -> bool:
        is_available = True
        for i in range(len(self.__list_of_courses)):
            if self.__list_of_courses[i].equals(checked_course):
                # if it finds the course then checks whether it
                # is retake-able or not then it returns false or true immediately
                is_available = self.__list_of_grades[i].is_retakable_grade()
                break

        return is_available

    # returns lists in form of string from transcript
    def get_student_transcript_string_list(self) -> list[str]:
        transcript_string = [self.__str__()]
        for i in range(len(self.__list_of_courses)):
            transcript_string.append(str(i+1) + "-)" + self.__list_of_courses[i - 1].__str__() + " " +
                                     self.__list_of_grades[i - 1].__str__())

        return transcript_string

    # override str func
    def __str__(self):
        digit_precision = 2
        temp_gpa = round(self.__gpa, digit_precision)
        out_string = "Cumulative Gpa: " + str(temp_gpa)
        out_string += "\nCumulative Credits: " + str(self.__student_credits)
        out_string += "\nCurrent Year: " + str(self.__student_year)
        return out_string

    # calculates the credits from taken courses
    def __calculate_credit(self):
        if self.__list_of_courses is None:
            logging.warning("No courses for student found!")
            return 0
        elif len(self.__list_of_courses) == 0:
            logging.warning("No courses for student found!")
        if self.__list_of_grades is None:
            logging.warning("No grades for student found!")
            return 0
        elif len(self.__list_of_grades) == 0:
            logging.warning("No grades for student found!")

        for i in range(len(self.__list_of_courses)):
            if self.__list_of_grades[i].get_numerical_grade() >= 35:
                self.__student_credits += self.__list_of_courses[i].get_course_credit()

    # calculates the gpa from grades
    def __calculate_gpa(self):
        total_credit = 0
        if self.__list_of_courses is None:
            logging.warning("No courses for student found!")
            return 0
        elif len(self.__list_of_courses) == 0:
            logging.warning("No courses for student found!")
        if self.__list_of_grades is None:
            logging.warning("No courses for student found!")
        elif len(self.__list_of_grades) == 0:
            logging.warning("No courses for student found!")

        weighted_values = []
        for i in range(len(self.__list_of_grades)):
            credit = self.__list_of_courses[i].get_course_credit()
            grade = self.__list_of_grades[i].get_numerical_grade()
            weighted_values.append(credit * grade)
            total_credit += credit

        total_weighted_values = 0

        for i in range(len(weighted_values)):
            total_weighted_values += weighted_values[i]

        self.__gpa = float(total_weighted_values) / (float(total_credit) * 25)

    # calculates the year from credits and gpa
    def __calculate_year(self):
        if self.__student_credits > 90:
            self.__student_year = 4
        elif self.__student_credits > 60:
            self.__student_year = 3
        elif self.__student_credits > 30:
            self.__student_year = 2

        if self.__student_year <= 3 <= self.__gpa:
            self.__student_year += 1

    # getters
    def get_gpa(self):
        return self.__gpa

    def get_student_year(self):
        return self.__student_year

    def get_student_credit(self):
        return self.__student_credits

    def get_list_of_grades(self):
        return self.__list_of_grades

    def get_list_of_courses(self):
        return self.__list_of_courses
